File: aws-ec2/operation/register_instances_with_classic_load_balancer.rb
=begin
##########################################################################
#
#  INFIVERVE TECHNOLOGIES PTE LIMITED CONFIDENTIAL
#  __________________
# 
#  (C) INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE
#  All Rights Reserved.
#  Product / Project: Flint IT Automation Platform
#  NOTICE:  All information contained herein is, and remains
#  the property of INFIVERVE TECHNOLOGIES PTE LIMITED.
#  The intellectual and technical concepts contained
#  herein are proprietary to INFIVERVE TECHNOLOGIES PTE LIMITED.
#  Dissemination of this information or any form of reproduction of this material
#  is strictly forbidden unless prior written permission is obtained
#  from INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE.
=end

@log.trace("Started executing 'fb-cloud:aws-ec2:operation:register_instances_with_classic_load_balancer.rb' flintbit...")
begin
connector_name = @input.get('connector_name')	      # Name of the Amazon EC2 Connector
load_balancer_name = @input.get('name') # Load Balancer Name
instances_to_register = @input.get('instances') # Instances to register to load balancer
action = 'register-instance-with-classic-load-balancer'
@access_key = @input.get('access-key')
@secret_key = @input.get('security-key')
request_timeout = @input.get('timeout')# Execution time of the Flintbit in milliseconds (default timeout is 60000 milloseconds)
region = @input.get('region')

if !load_balancer_name.nil? && !load_balancer_name.empty?
			if !instances_to_register.nil? && !instances_to_register.empty?
				connector_call = @call.connector(connector_name)
										.set('action',action)
										.set('load-balancer-name',load_balancer_name)
										.set('instance-id-list',instances_to_register)
										.set('access-key', @access_key)
            		    .set('security-key', @secret_key)

										if !request_timeout.nil? && !request_timeout.empty?
		            			connector_call.timeout(request_timeout)
		            		else
		            			@log.trace("Calling #{connector_name} with default timeout...")
		            		end
										if !region.nil? && !region.empty?
		            	         response = connector_call.set('region', region).sync
	            	    else
	            	    	response = connector_call.sync
	            	        @log.trace("region is not provided so using default region 'us-east-1'")
	            	    end
			else
				raise "Please provide instances to register"
			end
else
	raise "Please provide load balancer name"
end

if response.exitcode == 0
    @log.info("SUCCESS in executing #{connector_name} where, exitcode : #{response.exitcode} | message : #{response.message}")
    @output.set('message', response.message).set('exit-code', 0)
else
    @log.error("ERROR in executing #{connector_name} where, exitcode : #{response.exitcode} | message : #{response.message}")
		response=response.to_s
		if !response.empty?
		@output.set('message', response_message).set('exit-code', 1).setraw('error-details',response.to_s)
		else
		@output.set('message', response_message).set('exit-code', 1)
		end
end

rescue Exception => e
    @log.error(e.message)
    @output.set('exit-code', 1).set('message', e.message)
end

@log.trace("Finished executing 'fb-cloud:aws-ec2:operation:register_instances_with_classic_load_balancer.rb' flintbit")
# end

File: aws-ec2/operation/workflow/register_instances_with_classic_load_balancer.js
/*************************************************************************
 * 
 * INFIVERVE TECHNOLOGIES PTE LIMITED CONFIDENTIAL
 * __________________
 * 
 * (C) INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE
 * All Rights Reserved.
 * Product / Project: Flint IT Automation Platform
 * NOTICE:  All information contained herein is, and remains
 * the property of INFIVERVE TECHNOLOGIES PTE LIMITED.
 * The intellectual and technical concepts contained
 * herein are proprietary to INFIVERVE TECHNOLOGIES PTE LIMITED.
 * Dissemination of this information or any form of reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE.
 */

log.trace("Started executing the flintbit 'fb-cloud:aws-ec2:operation:workflow:register_instances_with_classic_load_balancer.js'");


input_scope = JSON.parse(input);   //For checking if the input contains the necessary keys

//Connector Name  - mandatory
connector_name = "amazon-ec2";
connector_call = call.connector(connector_name);
log.info("Connector Name: "+connector_name);

//Action - mandatory
action = "register-instance-with-classic-load-balancer";
log.info("Action: "+action);

connector_call.set("action",action);

if(input_scope.hasOwnProperty("cloud_connection")){

    //Access Key - mandatory
    access_key = input_scope.cloud_connection.encryptedCredentials["access_key"];
    if(access_key!=null || access_key!=""){
       connector_call.set("access-key",access_key); 
       log.info("Access Key is given");
    }
    else{
        log.error("Access-Key is null or empty string.");
    }

    //Security Key - mandatory
    secret_key = input_scope.cloud_connection.encryptedCredentials["secret_key"];
    if(secret_key!=null || secret_key!=""){
       connector_call.set("security-key",secret_key); 
       log.info("Security Key is given");
    }
    else{
        log.error("Security-Key is null or empty string.");
    }

    //Load Balancer Name - mandatory
    if(input_scope.hasOwnProperty("load_balancer_name")){
        load_balancer_name = input.get("load_balancer_name");
        if(load_balancer_name!=null || load_balancer_name!=""){
            connector_call.set("load-balancer-name",load_balancer_name); 
            log.info("Load Balancer Name: "+load_balancer_name);
        }
        else{
            log.error("load_balancer_name is null or empty string.");
        }
    }
    else{  //load_balancer_name key not present in input JSON 
        log.error("Input does not contain the key 'load_balancer_name'");
    }

    //Instances To Register - mandatory
    if(input_scope.hasOwnProperty("instances_to_register")){
        instances_to_register = input.get("instances_to_register");
        if(instances_to_register!=null || instances_to_register!=""){
            
            index = instances_to_register.indexOf(",");
            if(index!=-1){
                instances_to_register = instances_to_register.split(",");
                connector_call.set("instance-id-list",instances_to_register); 
                log.info("Instances to Register: "+instances_to_register);
            }
            else{
                instances_array = [];
                instances_array.push(instances_to_register);
                connector_call.set("instance-id-list",instances_array); 
                log.info("Instances to Register: "+instances_array);
            }
        }
        else{
            log.error("instances_to_register is null or empty string.");
        }
    }
    else{  //instances_to_register key not present in input JSON 
        log.error("Input does not contain the key 'instances_to_register'");
    }

    //Region - mandatory
    if(input_scope.hasOwnProperty("region")){
        region = input.get("region");
        if(region!=null || region!=""){
            connector_call.set("region",region); 
            log.info("Region: "+region);
        }
        else{
            log.error("Region is null or empty string.");
        }
    }
    else{  //region key not present in input JSON 
        log.error("Input does not contain the key 'region'");
    }

    //Timeout - NOT mandatory
    if(input_scope.hasOwnProperty("request_timeout")){
        request_timeout = input.get("request_timeout");
        if(request_timeout!=null || request_timeout!=""){
            connector_call.set("timeout",request_timeout); 
            log.info("Request Timeout: "+request_timeout);
        }
        else{
            connector_call.set("timeout",240000); 
            log.info("request_timeout not given. Setting 240000 miliseconds as timeout");
        }
    }
    else{
        connector_call.set("timeout",240000); 
        log.info("request_timeout not given. Setting 240000 miliseconds as timeout");
    }

    //Connector Call
    response = connector_call.sync();

    //Response Metaparameters
    response_exitcode = response.exitcode();
    response_message = response.message();

    if(response_exitcode==0){
        output.set("exit-code",response_exitcode)
                .set("message",response_message)
                .set("user_message","Instances Registered Successfully.");
        log.info("Instances Registered Successfully.");
        log.trace("Finished executing the flintbit 'fb-cloud:aws-ec2:operation:workflow:register_instances_with_classic_load_balancer.js' successfully");
    }
    else{
        log.error("Failure in execution, message:"+response_message+" | exitcode:"+response_exitcode);
        output.set("error",response_message).set("exit-code",-1);
        log.trace("finished executing 'fb-cloud:aws-ec2:operation:workflow:register_instances_with_classic_load_balancer.js' with errors")
    }
}
else{   //cloud_connection key not present in input JSON
    log.error("Cloud Connection not given. Can not authenticate without Secret-Key and Access-Key");
}
